### ISSUE TITLE
Resizing rows in dynamic list

### ISSUE THREAD

**Issue by bebbi** [tannerlinsley/react-virtual#28]
I've given the dynamic list of this great lib a spin and encountered two issues:

1. My rows can expand/collapse on click (i.e. change height), but the virtualized list doesn't adapt the item positions. Could this benefit from a `ResizeObserver` hook such as [useMeasure](https://github.com/streamich/react-use/blob/master/docs/useMeasure.md)?
2. It would be very nice to have a mechanism where the dynamic list's overall initial height is set such that it doesn't cut off partial rows. See e.g. the first example shown in [react-collapse](https://www.npmjs.com/package/react-collapse) where the height limit is set by the number of visible paragraphs.

**Comment by piecyk** [comment]
Yes, using `ResizeObserver` is the way to go. Current it's possible to implement this with small row wrapper, checkout demo at 
https://codesandbox.io/s/resizeobserver-react-virtual-6131c?file=/src/index.js

IMHO as ResizeObserver is supported in every modern browser https://caniuse.com/resizeobserver maybe we can officially supporting it.

**Comment by gitcatrat** [comment]
@piecyk Have you implemented resize observer it with image content? I.e image loads after some delay.

Scrolling down seems to work but scrolling up is super janky because old accurate height value is thrown away and replaced by the height again where image is missing/is loading again.

**Comment by kidroca** [comment]
I've found that for any dynamic size to be able to work we shouldn't set a fixed `height` on the row elements 
I also have content that allows to expand an item and it doesn't work unless I substitute `height` for `min-height` 

```js
           {rowVirtualizer.virtualItems.map(virtualRow => (
             <div
               key={virtualRow.index}
               ref={virtualRow.measureRef}
               className={virtualRow.index % 2 ? 'ListItemOdd' : 'ListItemEven'}
               style={{
                 position: 'absolute',
                 top: 0,
                 left: 0,
                 width: '100%',
                 // This change
                 minHeight: `${virtualRow.size}px`,
                 transform: `translateY(${virtualRow.start}px)`,
               }}
             >
               Row {virtualRow.index}
             </div>
           ))}
```

For example my inner content would expand an item from 100px to 350px, but it won't work when the row has a fixed height of 100px 
When I use `min-height` the row is allowed to expand, which in turn gets captured by `measureRef` and the change propagates to `virtualRow.size` and `virtualRow.start` 

I imagine to be able to both expand and shrink we should not be setting any height at all, but only `measureRef`

**Comment by alvis** [comment]
Thanks @kidroca. Seems like your approach is a better one than using `ResizeObserver`. Didn't think of that.

**Comment by piecyk** [comment]
Yes, for dynamic case you should not set height or min-height, the size will be read via measureRef, the examples now are bit misleading. Keep in mind that if the item will change it size without `ResizeObserver` this change need to trigger re-render, and measureRef need to be called ( for case when visible items will not change, hack with `ref={el => virtualRow.measureRef(el) }` is needed ) 

Checkout the example from https://github.com/tannerlinsley/react-virtual/discussions/212#discussioncomment-1587478 regarding ResizeObserver and stale measureRef for performance optimisation.

**Comment by gitcatrat** [comment]
`ResizeObserver` is a must if your list items contain images, etc elements and data that change the height asynchronously, unless you know their heights before they load which is rarely the case (could save image size to database on resize).

Unfortunately there are few issues that examples don't address and that I haven't figured out myself as well:
* **performance** - attaching and detaching observers can be expensive, especially if you drag the scrollbar
* **scroll direction** 
  * scrolling down works because loaded list items push lower items down, it's usually not visible
  * but if you're scrolling up and upper items heights change then all on-screen items are pushed down

If anyone has any ideas, I'm all ears. If I manage to find a good solution, I'll let you know as well.

**Comment by piecyk** [comment]
@gitcatrat did you check the `ResizeObserver` example from https://github.com/tannerlinsley/react-virtual/discussions/212#discussioncomment-1587478 using stale measureRef, performance is good, imho this is best what we can have now.

**Comment by ISilviu** [comment]
> Yes, for dynamic case you should not set height or min-height, the size will be read via measureRef, the examples now are bit misleading. Keep in mind that if the item will change it size without `ResizeObserver` this change need to trigger re-render, and measureRef need to be called ( for case when visible items will not change, hack with `ref={el => virtualRow.measureRef(el) }` is needed )
> 
> Checkout the example from [#212 (reply in thread)](https://github.com/tannerlinsley/react-virtual/discussions/212#discussioncomment-1587478) regarding ResizeObserver and stale measureRef for performance optimisation.

I have used min-height to implement a expand feature, yet the performance seems a bit off. 
Regardless of that, I'm having issues with the scroll. Whenever I expand/collapse a row, the list scrolls a bit down, hiding my content. Right now I'm trying to use something like `React.useEffect(() => scrollToIndex(clickedItemIndex), [clickedItemIndex]);` to ensure my scroll stays in place, and it does the job in some cases. However, I was wondering if you have any suggestion; i.e. should I try to use ResizeObserver?

Unfortunately, I don't have a codesandbox to illustrate this.

**Comment by AgarwalPragy** [comment]
@tannerlinsley since v3 is out, are there any plans to provide @piecyk's solution as the default behavior?
https://github.com/TanStack/virtual/issues/28#issuecomment-817182500

If there aren't any significant performance implications, then offering that as a default will be good, since elements resizing after render is a common usecase.

**Comment by alissaVrk** [comment]
here is a possible solution

the trick is to call `measureElement` with an object that you can extract the item new height from and not the element itself

https://codesandbox.io/s/scroll-resize-working-wr1t78?file=/src/App.tsx

**Comment by nconfrey** [comment]
For anyone else hitting the issues of scrolling up causing the list to stutter and resize, this code inside of the `measureElement` of the Virtualizer helped:
```
measureElement: (element, entry, instance) => {
      const direction = instance.scrollDirection
      if (direction === "forward" || direction === null) {
        return element.scrollHeight
      } else {
        // don't remeasure if we are scrolling up
        const indexKey = Number(element.getAttribute("data-index"))
        let cacheMeasurement = instance.itemSizeCache.get(indexKey)
        return cacheMeasurement
      }
    }
```

**Comment by chien-le-tpv-clv** [comment]
> I've found that for any dynamic size to be able to work we shouldn't set a fixed `height` on the row elements I also have content that allows to expand an item and it doesn't work unless I substitute `height` for `min-height`
> 
>            {rowVirtualizer.virtualItems.map(virtualRow => (
>              <div
>                key={virtualRow.index}
>                ref={virtualRow.measureRef}
>                className={virtualRow.index % 2 ? 'ListItemOdd' : 'ListItemEven'}
>                style={{
>                  position: 'absolute',
>                  top: 0,
>                  left: 0,
>                  width: '100%',
>                  // This change
>                  minHeight: `${virtualRow.size}px`,
>                  transform: `translateY(${virtualRow.start}px)`,
>                }}
>              >
>                Row {virtualRow.index}
>              </div>
>            ))}
> For example my inner content would expand an item from 100px to 350px, but it won't work when the row has a fixed height of 100px When I use `min-height` the row is allowed to expand, which in turn gets captured by `measureRef` and the change propagates to `virtualRow.size` and `virtualRow.start`
> 
> I imagine to be able to both expand and shrink we should not be setting any height at all, but only `measureRef`

add minHeight based on the virutalRow.size save me, thank you
